search-books/src/main/java/nix/education/service/TransactionalDataService.java

```diff
@@ -0,0 +1,52 @@
+package nix.education.service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import nix.education.entity.Book;
+import nix.education.entity.Tag;
+import nix.education.repository.BookRepository;
+import nix.education.repository.TagRepository;
+import nix.education.request.BookRequest;
+import org.springframework.stereotype.Service;
+
+@Service
+public class TransactionalDataService {
+
+    public Book getSavedBook(BookRequest bookRequest,
+                             BookRepository bookRepository,
+                             TagRepository tagRepository) {
+        Book book = new Book();
+        book.setAuthor(bookRequest.getAuthor());
+        book.setTitle(bookRequest.getTitle());
+        book.setDescription(bookRequest.getDescription());
+        book.setPublicationDate(bookRequest.getPublicationDate());
+        book.setTags(getResolvedTags(bookRequest.getTags(), tagRepository));
+        bookRepository.save(book);
+        book.setTagTitles(getTagTitles(book));
+        return book;
+    }
+
+    public List<String> getTagTitles(Book book) {
+        return book.getTags().stream().map(Tag::getTag).collect(Collectors.toList());
+    }
+
+    public List<Book> getBooksByIds(List<Long> ids, BookRepository bookRepository) {
+        List<Book> books = new ArrayList<>();
+        ids.forEach(id -> books.add(bookRepository.findById(id).orElseThrow()));
+        return books;
+    }
+
+    public List<Tag> getResolvedTags(List<String> tags, TagRepository tagRepository) {
+        List<Tag> allTags = (List<Tag>) tagRepository.findAll();
```

Comment:
I thought I didn't save my comment I left yesterday... Do we nee to fetch all the tags? What if there are a lot of them? And we have to load all of them to the memory.

Reply:
I've already changed this logic and now it works another.